internal/manifest/ready_check.go

```diff
@@ -232,3 +72,31 @@ func isDeploymentReady(clt declarativev2.Client, resources []*resource.Info) boo
 	}
 	return false
 }
+
+func getPodsForDeployment(clt declarativev2.Client, deploy *apiappsv1.Deployment) (*apicorev1.PodList, error) {
+	podList := &apicorev1.PodList{}
+	listOptions := &client.ListOptions{
+		Namespace:     deploy.Namespace,
+		LabelSelector: k8slabels.SelectorFromSet(deploy.Spec.Selector.MatchLabels),
+	}
+	if err := clt.List(context.TODO(), podList, listOptions); err != nil {
+		return nil, fmt.Errorf("failed to list pods: %w", err)
+	}
+	return podList, nil
+}
+
+func getPodsState(podList *apicorev1.PodList) shared.State {
+	for _, pod := range podList.Items {
+		for _, condition := range pod.Status.ContainerStatuses {
+			switch {
+			case *condition.Started && condition.Ready:
```

Comment:
why not use containerStates? https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-states

`ContainerStatus.State` contains three state, which can be mapped to our shared state.

Reply:
Because I had the Problem when using ContainerStatus, then it was flickering between `Waiting` and `Terminated`. When a container gets terminated, the container will be restarted and thus have again the status `Waiting` since it tries to fetch the image. If for example the image does not exist (ImagePullBackOff) or the image crashes it terminates again and restart. This is reflected in the restart count.

Using the conditions is in my opinion the better way to map the current behaviour of the deployment to our states without any flickering.

Reply:
ok, then I'm fine with this `condition.Started`, however, it's an optional field, how about those `nil` case, I mean if the containerstatues doesn't have Started condition, your current implementation will have NPE.

You use two boolean variables (`*condition.Started` and `condition.Ready`) in the switch case, it should have 4 cases, but only have three in your implementation. The fourth case may never occured, but for consistent purpose, we should not ignore it. I suggest the rest two cases by default return `Error` state

```
case *condition.Started && condition.Ready:
  return shared.StateReady
case *condition.Started && !condition.Ready:
  return shared.StateProcessing
default:
  return shared.StateError
```

and please write a unit test for this `getPodsState` function.